components/gfx/text/text_run.rs

```diff
@@ -254,26 +264,28 @@ impl<'a> TextRun {
                 continue;
             }
             if !slice.is_empty() {
-                glyphs.push(GlyphRun {
-                    glyph_store: font.shape_text(&text[slice.clone()], options),
-                    range: Range::new(
-                        ByteIndex(slice.start as isize),
-                        ByteIndex(slice.len() as isize),
-                    ),
-                });
+                push_range(&slice, options);
             }
             if !whitespace.is_empty() {
                 let mut options = *options;
                 options
                     .flags
                     .insert(ShapingFlags::IS_WHITESPACE_SHAPING_FLAG);
-                glyphs.push(GlyphRun {
-                    glyph_store: font.shape_text(&text[whitespace.clone()], &options),
-                    range: Range::new(
-                        ByteIndex(whitespace.start as isize),
-                        ByteIndex(whitespace.len() as isize),
-                    ),
-                });
+
+                // The breaker breaks after every newline, so either there is none,
+                // or there is exactly one at the very end. In the latter case,
+                // split it into a different run. That's because shaping considers
+                // a newline to have the same advance as a space, but during layout
+                // we want to treat the newline as having no advance.
+                if word.chars().last().unwrap() == '\n' {
```

Comment:
Is there any way to avoid this call to `chars()` here? It has to walk the entire string again and separate characters into Unicode characters. Since this code is somewhat performance sensitive, it would be good to avoid it.

Reply:
Is `as_bytes()` better?